### PR TITLE
[13.0][IMP] sale_order_product_recommendation: Exclude delivery carrier products from recommendations

### PR DESCRIPTION
**Pull request by sergio-teruel**
cc @Tecnativa

ping @carlosdauden @CarlosRoca13 @LudLaf 

TT35965

### CODE REVIEW

**Review by carlosdauden — APPROVED** [review]

**Review by pedrobaeza — COMMENTED** [review]
Why not directly exclude SO lines with `is_delivery` field set to True?

**Comment by sergio-teruel** [comment]
@pedrobaeza Changes done!!

File: sale_order_product_recommendation/wizards/sale_order_recommendation.py [hunk 1]
```diff
@@ -75,12 +75,16 @@ def _recomendable_sale_order_lines_domain(self):
...
         )
-        return [
+        domain = [
             ("order_id", "in", (other_sales - self.order_id).ids),
             ("product_id.active", "=", True),
             ("product_id.sale_ok", "=", True),
             ("qty_delivered", "!=", 0.0),
         ]
+        # Exclude delivery products
+        if "delivery.carrier" in self.env.registry:
```
Comment (pedrobaeza) [inline sale_order_product_recommendation/wizards/sale_order_recommendation.py]:
```suggestion
        if "is_delivery" in self.env["sale.order.line"]._fields:
```

File: sale_order_product_recommendation/wizards/sale_order_recommendation.py [hunk 2]
```diff
@@ -128,8 +132,9 @@ def _generate_recommendations(self):
         found_dict = {l["product_id"][0]: l for l in found_lines}
         recommendation_lines = self.env["sale.order.recommendation.line"]
         existing_product_ids = set()
-        # Always recommend all products already present in the linked SO
-        for line in self.order_id.order_line:
+        # Always recommend all products already present in the linked SO except delivery
+        # carrier products
+        for line in self.order_id.order_line.filtered(lambda ln: not ln.is_delivery):
```
Comment (pedrobaeza) [inline sale_order_product_recommendation/wizards/sale_order_recommendation.py]:
```suggestion
        lines = self.order_id.order_line
        if "is_delivery" in self.env["sale.order.line"]._fields:
            lines = lines.filtered(lambda ln: not ln.is_delivery)
        for line in lines:
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by rousseldenis** [comment]
Mmmh, don't like code based on unloaded things. The best is to do a glue module...

**Comment by pedrobaeza** [comment]
Well, it's a good trade-off. We know that it's not the perfect one, but have a glue module `*_delivery` for every module we have... Did you contribute in Odoo a hook method for excluding SO lines that is hooked by `delivery` module? I remember something like that. That can be a good alternative.

**Comment by sergio-teruel** [comment]
Hi
We can not use the method _is_delivery() from sale module because we are doing a domain for a readgroup query.

File: sale_order_product_recommendation/wizards/sale_order_recommendation.py (hunk 2 above)
Comment (pedrobaeza) [reply sale_order_product_recommendation/wizards/sale_order_recommendation.py]: This one is missing

Comment (sergio-teruel) [reply sale_order_product_recommendation/wizards/sale_order_recommendation.py]: Here we can use method _is_delivery() from sale module

**Comment by sergio-teruel** [comment]
Changes done!

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**